File: api_java_distributed_order_platform/src/main/java/com/project/order/application/usecase/payment/CreatePaymentUseCase.java
package com.project.order.application.usecase.payment;

import com.project.order.application.exception.BusinessRuleException;
import com.project.order.application.exception.NotFoundException;
import com.project.order.application.port.EventPublisherRepository;
import com.project.order.application.port.OrderRepository;
import com.project.order.application.port.PaymentGatewayRepository;
import com.project.order.application.port.PaymentRepository;
import com.project.order.application.port.UnitOfWorkRepository;
import com.project.order.domain.model.Order;
import com.project.order.domain.model.OrderStatus;
import com.project.order.domain.model.Payment;
import com.project.order.domain.model.Stock;
import com.project.order.shared.interfaces.GatewayResult;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

/**
 * Igual ao CreatePaymentUseCase do NestJS: a chamada ao gateway acontece
 * FORA de qualquer transação de banco (I/O externo não deve segurar uma
 * conexão/transação aberta). Por isso aqui a gente usa o UnitOfWorkRepository
 * manual (begin/commit/rollback) só depois que o resultado do gateway já
 * voltou, em vez de um único método @Transactional cobrindo tudo.
 */
@Service
public class CreatePaymentUseCase {

    private static final Logger log = LoggerFactory.getLogger(CreatePaymentUseCase.class);
    private static final int MAX_RETRIES = 3;

    private final PaymentRepository paymentRepository;
    private final OrderRepository orderRepository;
    private final UnitOfWorkRepository unitOfWork;
    private final PaymentGatewayRepository gateway;
    private final EventPublisherRepository eventPublisher;

    public CreatePaymentUseCase(
            PaymentRepository paymentRepository,
            OrderRepository orderRepository,
            UnitOfWorkRepository unitOfWork,
            PaymentGatewayRepository gateway,
            EventPublisherRepository eventPublisher
    ) {
        this.paymentRepository = paymentRepository;
        this.orderRepository = orderRepository;
        this.unitOfWork = unitOfWork;
        this.gateway = gateway;
        this.eventPublisher = eventPublisher;
    }

    public Payment execute(String orderId, Double requestedAmount) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new NotFoundException("Order not found"));
        if (order.getStatus() != OrderStatus.PENDING) {
            throw new BusinessRuleException("Order is not pending");
        }
        if (order.getItems().isEmpty()) {
            throw new BusinessRuleException("Order has no items");
        }
        if (requestedAmount != null && !requestedAmount.equals(order.getTotal().getValue())) {
            throw new BusinessRuleException("Payment amount does not match order total");
        }

        Payment payment = paymentRepository.create(Payment.create(order.getId(), order.getTotal().getValue()));
        String idempotencyKey = payment.getId();

        GatewayResult gatewayResult = chargeWithRetry(payment, idempotencyKey);

        unitOfWork.begin();
        try {
            if (gatewayResult.success()) {
                for (var item : order.getItems()) {
                    Stock stock = unitOfWork.getStockRepository().findById(item.getProductId())
                            .orElseThrow(() -> new NotFoundException("Stock not found for product " + item.getProductId()));
                    stock.consume(item.getQuantity());
                    unitOfWork.getStockRepository().update(stock);
                    publishQuietly(() -> eventPublisher.publishStockConsumed(stock), "StockConsumed", stock.getProductId());
                }

                payment.markAsPaid(
                        gatewayResult.externalId() != null ? gatewayResult.externalId() : idempotencyKey,
                        gatewayResult.raw()
                );
                order.confirmOrder();
            } else {
                String code = gatewayResult.errorCode() != null ? gatewayResult.errorCode() : "GATEWAY_ERROR";
                String message = gatewayResult.errorMessage() != null ? gatewayResult.errorMessage() : "Payment failed";

                payment.markAsFailed(code, message, gatewayResult.raw());
                order.registerPaymentFailure(code, message);

                for (var item : order.getItems()) {
                    Stock stock = unitOfWork.getStockRepository().findById(item.getProductId())
                            .orElseThrow(() -> new NotFoundException("Stock not found for product " + item.getProductId()));
                    stock.release(item.getQuantity());
                    unitOfWork.getStockRepository().update(stock);
                    publishQuietly(() -> eventPublisher.publishStockReleased(stock), "StockReleased", stock.getProductId());
                }
            }

            unitOfWork.getPaymentRepository().update(payment);
            unitOfWork.getOrderRepository().update(order);
            unitOfWork.commit();
        } catch (RuntimeException e) {
            safeRollback();
            publishQuietly(() -> eventPublisher.publishPaymentFailed(payment), "PaymentFailed", payment.getId());
            throw e;
        }

        if (gatewayResult.success()) {
            publishQuietly(() -> eventPublisher.publishPaymentPaid(payment), "PaymentPaid", payment.getId());
        } else {
            publishQuietly(() -> eventPublisher.publishPaymentFailed(payment), "PaymentFailed", payment.getId());
        }

        return payment;
    }

    private GatewayResult chargeWithRetry(Payment payment, String idempotencyKey) {
        try {
            GatewayResult result = gateway.createCharge(payment.getId(), payment.getOrderId(), payment.getAmount().getValue(), idempotencyKey);
            if (result.success() || "BUSINESS".equals(result.kind())) {
                return result;
            }
            return retryRetrieve(idempotencyKey);
        } catch (RuntimeException e) {
            if (!isRetriable(e)) {
                return new GatewayResult(false, null, "failed", "PAYMENT_REJECTED", e.getMessage(), "BUSINESS", null);
            }
            return retryRetrieve(idempotencyKey);
        }
    }

    private GatewayResult retryRetrieve(String idempotencyKey) {
        GatewayResult lastResult = new GatewayResult(false, null, "pending", "GATEWAY_UNAVAILABLE", "Gateway unavailable after retries", "RETRIABLE", null);

        for (int attempt = 1; attempt <= MAX_RETRIES; attempt++) {
            sleep(attempt * 500L);
            try {
                lastResult = gateway.retrieveCharge(idempotencyKey, null);
                if (lastResult.success() || "BUSINESS".equals(lastResult.kind())) {
                    return lastResult;
                }
            } catch (RuntimeException e) {
                boolean retriable = isRetriable(e);
                lastResult = new GatewayResult(false, null, "failed",
                        retriable ? "GATEWAY_UNAVAILABLE" : "PAYMENT_REJECTED",
                        e.getMessage(), retriable ? "RETRIABLE" : "BUSINESS", null);
                if (!retriable) {
                    return lastResult;
                }
            }
        }
        return lastResult;
    }

    private boolean isRetriable(RuntimeException e) {
        String message = e.getMessage() == null ? "" : e.getMessage().toLowerCase();
        return message.contains("etimedout") || message.contains("econnreset") || message.contains("enotfound")
                || message.contains("timeout") || message.contains("socket") || message.contains("network")
                || message.contains("502") || message.contains("503") || message.contains("504");
    }

    private void safeRollback() {
        try {
            unitOfWork.rollback();
        } catch (Exception ignored) {
            // ponytail: melhor esforço, transação já pode ter sido finalizada
        }
    }

    private void sleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    private void publishQuietly(Runnable action, String eventName, String subjectId) {
        try {
            action.run();
        } catch (Exception e) {
            log.error("Failed to publish {} for {}", eventName, subjectId, e);
        }
    }
}

File: api_nestjs_distributed_order_platform/src/application/use-case/payment/create-payment.use-case.ts
import {
    BadRequestException,
    Inject,
    Injectable,
    Logger,
    NotFoundException,
} from "@nestjs/common";
import type { EventPublisherRepositoryInterface } from "src/application/port/event-publisher.repository";
import type { OrderRepositoryInterface } from "src/application/port/order.repository";
import type { PaymentGatewayInterface } from "src/application/port/payment-gateway.repository";
import type { PaymentRepositoryInterface } from "src/application/port/payment.repository";
import type { UnitOfWorkInterface } from "src/application/port/unit-of-work.respository";
import { PaymentEntity } from "src/domain/entities/payment.entity";
import { StockEntity } from "src/domain/entities/stock.entity";
import { StatusOrder } from "src/shared/enums/status-order.enum";
import type { GatewayResult } from "src/shared/interfaces/gateway.interface";
import type { PaymentCreateAttributes } from "src/shared/interfaces/payment.interface";
import { GATEWAY_REPOSITORY } from "src/shared/tokens_nest/gateway.token";
import { ORDER_REPOSITORY } from "src/shared/tokens_nest/order.token";
import {
    PAYMENT_REPOSITORY,
    UNIT_OF_WORK_REPOSITORY,
} from "src/shared/tokens_nest/payment.token";
import { EVENT_PUBLISHER } from "src/shared/tokens_nest/rabbitmq.token";

const MAX_RETRIES = 3;

@Injectable()
export class CreatePaymentUseCase {
    private readonly logger = new Logger(CreatePaymentUseCase.name);

    constructor(
        @Inject(PAYMENT_REPOSITORY)
        private readonly paymentRepository: PaymentRepositoryInterface,
        @Inject(ORDER_REPOSITORY)
        private readonly orderRepository: OrderRepositoryInterface,
        @Inject(UNIT_OF_WORK_REPOSITORY)
        private readonly uow: UnitOfWorkInterface,
        @Inject(GATEWAY_REPOSITORY)
        private readonly gateway: PaymentGatewayInterface,
        @Inject(EVENT_PUBLISHER)
        private readonly eventPublisher: EventPublisherRepositoryInterface,
    ) {}

    async execute(input: PaymentCreateAttributes): Promise<PaymentEntity> {
        const order = await this.orderRepository.findById(input.order_id);
        if (!order) {
            throw new NotFoundException('Order not found');
        }
        if (order.getStatus() !== StatusOrder.PENDING) {
            throw new BadRequestException('Order is not pending');
        }
        if (order.getItems().length === 0) {
            throw new BadRequestException('Order has no items');
        }
        if (
            input.amount !== undefined &&
            Number(input.amount) !== order.getTotal()
        ) {
            throw new BadRequestException(
                'Payment amount does not match order total',
            );
        }

        const payment = await this.paymentRepository.createPayment(
            new PaymentEntity(order.getId(), order.getTotal()),
        );
        const idempotencyKey = payment.getId();

        // Gateway FORA da TX — estoque permanece reservado até o resultado
        const gatewayResult = await this.chargeWithRetry(payment, idempotencyKey);

        await this.uow.begin();
        try {
            const tx = this.uow.getTransaction()!;

            if (gatewayResult.success) {
                for (const item of order.getItems()) {
                    const stock = await this.uow
                        .getStockRepository()
                        .findByProductId(item.getProductId(), tx);
                    if (!stock) {
                        throw new NotFoundException(
                            `Stock not found for product ${item.getProductId()}`,
                        );
                    }
                    stock.consume(item.getQuantity());
                    await this.uow
                        .getStockRepository()
                        .update(stock, undefined, tx);

                    try{
                        await this.eventPublisher.publishStockConsumed(stock);
                    }catch(error){
                        this.logger.error(
                            `Failed to publish StockConsumed for product ${item.getProductId()}`,
                            error instanceof Error ? error.stack : String(error),
                        );
                    }
                }

                payment.markAsPaid(
                    gatewayResult.externalId ?? idempotencyKey,
                    gatewayResult.raw,
                );
                order.confirmOrder();
                
            } else {
                const code = gatewayResult.errorCode ?? 'GATEWAY_ERROR';
                const message =
                    gatewayResult.errorMessage ?? 'Payment failed';

                payment.markAsFailed(code, message, gatewayResult.raw);
                order.registerPaymentFailure(code, message);

                for (const item of order.getItems()) {
                    const stock = await this.uow
                        .getStockRepository()
                        .release(item.getProductId(), item.getQuantity(), tx);
                    
                    try{
                        await this.eventPublisher.publishStockReleased(stock);
                    }catch(error){
                        this.logger.error(
                            `Failed to publish StockReleased for product ${item.getProductId()}`,
                            error instanceof Error ? error.stack : String(error),
                        );
                    }
                }
            }

            await this.uow.getPaymentRepository().updatePayment(payment, tx);
            await this.uow.getOrderRepository().update(order, tx);
            await this.uow.commit();
        } catch (error) {
            await this.safeRollback();
            try{
                await this.eventPublisher.publishPaymentFailed(payment);
            }catch(error){
                this.logger.error(
                    `Failed to publish PaymentFailed for payment ${payment.getId()}`,
                    error instanceof Error ? error.stack : String(error),
                );
            }
            throw error;
        }

        try{
            gatewayResult.success? await this.eventPublisher.publishPaymentPaid(payment) : await this.eventPublisher.publishPaymentFailed(payment);
        }catch(error){
            this.logger.error(
                `Failed to publish PaymentPaid for payment ${payment.getId()}`,
                error instanceof Error ? error.stack : String(error),
            );
        }
        
        return payment;
    }

    private async chargeWithRetry(
        payment: PaymentEntity,
        idempotencyKey: string,
    ): Promise<GatewayResult> {
        try {
            const result = await this.gateway.createCharge({
                paymentId: payment.getId(),
                orderId: payment.getOrderId(),
                amount: payment.getAmount(),
                idempotencyKey,
            });

            if (result.success || result.kind === 'BUSINESS') {
                return result;
            }

            return this.retryRetrieve(idempotencyKey);
        } catch (error) {
            if (!this.isRetriable(error)) {
                return {
                    success: false,
                    kind: 'BUSINESS',
                    status: 'failed',
                    errorCode: 'PAYMENT_REJECTED',
                    errorMessage:
                        error instanceof Error
                            ? error.message
                            : 'Payment rejected',
                    raw: error,
                };
            }

            return this.retryRetrieve(idempotencyKey);
        }
    }

    private async retryRetrieve(idempotencyKey: string): Promise<GatewayResult> {
        let lastResult: GatewayResult = {
            success: false,
            kind: 'RETRIABLE',
            status: 'pending',
            errorCode: 'GATEWAY_UNAVAILABLE',
            errorMessage: 'Gateway unavailable after retries',
        };

        for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
            await this.sleep(attempt * 500);

            try {
                lastResult = await this.gateway.retrieveCharge({
                    idempotencyKey,
                });

                if (lastResult.success || lastResult.kind === 'BUSINESS') {
                    return lastResult;
                }
            } catch (error) {
                lastResult = {
                    success: false,
                    kind: this.isRetriable(error) ? 'RETRIABLE' : 'BUSINESS',
                    status: 'failed',
                    errorCode: this.isRetriable(error)
                        ? 'GATEWAY_UNAVAILABLE'
                        : 'PAYMENT_REJECTED',
                    errorMessage:
                        error instanceof Error
                            ? error.message
                            : 'Gateway error',
                    raw: error,
                };

                if (!this.isRetriable(error)) {
                    return lastResult;
                }
            }
        }

        return lastResult;
    }

    private isRetriable(error: unknown): boolean {
        if (!(error instanceof Error)) {
            return true;
        }

        const message = error.message.toLowerCase();
        const retriableHints = [
            'etimedout',
            'econnreset',
            'enotfound',
            'timeout',
            'socket',
            'network',
            '502',
            '503',
            '504',
        ];

        return retriableHints.some((hint) => message.includes(hint));
    }

    private async safeRollback(): Promise<void> {
        try {
            if (this.uow.getTransaction()) {
                await this.uow.rollback();
            }
        } catch {
            // ignore
        }
    }

    private sleep(ms: number): Promise<void> {
        return new Promise((resolve) => setTimeout(resolve, ms));
    }
}
